wip temporary index encoding

rhu713 opened the pull request:
Release note (<category, see below>): <what> <show> <why>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71397)
<!-- Reviewable:end -->

Commit:
sql: add delete preserving index encoding, use for primary and secondary index

As part of the transition for Bulk Ops to use only MVCC operations, the index
backfiller must be rewritten. Part of the rewrite is to introduce a temporary
index at the beginning of the backfill encoded with a new delete-preserving
index encoding. This patch adds this encoding and uses the encoding for writes
to indices configured with this encoding.

Release note: None
